File: passgen_cli/core/generator.py
"""Password generation logic."""

import secrets
import random

from .constants import CHAR_SETS


def generate_password(length: int, char_types: list[str]) -> str:
    """Generate a password with specified character types and length."""
    all_chars = "".join(CHAR_SETS[ct] for ct in char_types if ct in CHAR_SETS)

    if not all_chars:
        raise ValueError("At least one character type must be selected")

    return "".join(secrets.choice(all_chars) for _ in range(length))


def get_random_char_types() -> list[str]:
    """Randomly select character types (at least 1, up to all 4)."""
    all_types = list(CHAR_SETS.keys())
    num_types = random.randint(1, len(all_types))
    return random.sample(all_types, num_types)

File: src/core/generator.ts
import crypto from "crypto";
import { CHAR_SETS, CharType } from "./constants.js";

export function generatePassword(length: number, charTypes: CharType[]): string {
  const allChars = charTypes
    .map((ct) => CHAR_SETS[ct])
    .join("");

  if (!allChars) {
    throw new Error("At least one character type must be selected");
  }

  let password = "";
  for (let i = 0; i < length; i++) {
    const randomIndex = crypto.randomInt(0, allChars.length);
    password += allChars[randomIndex];
  }
  return password;
}

export function getRandomCharTypes(): CharType[] {
  const allTypes: CharType[] = ["numbers", "lowercase", "uppercase", "special"];
  const numTypes = crypto.randomInt(1, allTypes.length + 1);

  // Fisher-Yates shuffle to select random types
  const shuffled = [...allTypes];
  for (let i = shuffled.length - 1; i > 0; i--) {
    const j = crypto.randomInt(0, i + 1);
    [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
  }

  return shuffled.slice(0, numTypes);
}
